<a href="https://colab.research.google.com/github/shreyankaul/scaler/blob/main/Delhivery.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Delhivery - Feature Engineering**

#Problem Statement

Delhivery is the largest and fastest-growing fully integrated player in India by revenue in Fiscal 2021. They aim to build the operating system for commerce, through a combination of world-class infrastructure, logistics operations of the highest quality, and cutting-edge engineering and technology capabilities.

The Data team builds intelligence and capabilities using this data that helps them to widen the gap between the quality, efficiency, and profitability of their business versus their competitors.

The company wants to understand and process the data coming out of data engineering pipelines:

• Clean, sanitize and manipulate data to get useful features out of raw fields

• Make sense out of the raw data and help the data science team to build forecasting models on it

#Importing Data and Python Libraries

In [3]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy.stats import ttest_rel

from sklearn.preprocessing import StandardScaler

In [5]:
!wget https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/001/551/original/delhivery_data.csv?1642751181 -O final.csv

--2026-05-26 13:59:47--  https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/001/551/original/delhivery_data.csv?1642751181
Resolving d2beiqkhq929f0.cloudfront.net (d2beiqkhq929f0.cloudfront.net)... 13.33.47.175, 13.33.47.29, 13.33.47.5, ...
Connecting to d2beiqkhq929f0.cloudfront.net (d2beiqkhq929f0.cloudfront.net)|13.33.47.175|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 55617130 (53M) [text/plain]
Saving to: ‘final.csv’

final.csv           100%[===================>]  53.04M  --.-KB/s    in 0.1s    

2026-05-26 13:59:47 (435 MB/s) - ‘final.csv’ saved [55617130/55617130]



In [6]:
df = pd.read_csv('/content/final.csv')

In [7]:
df.head()

,data,trip_creation_time,route_schedule_uuid,route_type,trip_uuid,source_center,source_name,destination_center,destination_name,od_start_time,...,cutoff_timestamp,actual_distance_to_destination,actual_time,osrm_time,osrm_distance,factor,segment_actual_time,segment_osrm_time,segment_osrm_distance,segment_factor
0,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,...,2018-09-20 04:27:55,10.435660,14.0,11.0,11.9653,1.272727,14.0,11.0,11.9653,1.272727
1,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,...,2018-09-20 04:17:55,18.936842,24.0,20.0,21.7243,1.200000,10.0,9.0,9.7590,1.111111
2,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,...,2018-09-20 04:01:19.505586,27.637279,40.0,28.0,32.5395,1.428571,16.0,7.0,10.8152,2.285714
3,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,...,2018-09-20 03:39:57,36.118028,62.0,40.0,45.5620,1.550000,21.0,12.0,13.0224,1.750000
4,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,...,2018-09-20 03:33:55,39.386040,68.0,44.0,54.2181,1.545455,6.0,5.0,3.9153,1.200000


In [8]:
df.shape

(144867, 24)

**Observation:**

The dataset contains 144867 rows and 24 columns, indicating a large shipment-level logistics dataset with multiple operational variables.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   data                            144867 non-null  object 
 1   trip_creation_time              144867 non-null  object 
 2   route_schedule_uuid             144867 non-null  object 
 3   route_type                      144867 non-null  object 
 4   trip_uuid                       144867 non-null  object 
 5   source_center                   144867 non-null  object 
 6   source_name                     144574 non-null  object 
 7   destination_center              144867 non-null  object 
 8   destination_name                144606 non-null  object 
 9   od_start_time                   144867 non-null  object 
 10  od_end_time                     144867 non-null  object 
 11  start_scan_to_end_scan          144867 non-null  float64
 12  is_cutoff       

**Observation:**

The dataset contains 144867 records and 24 columns consisting of categorical, numerical, boolean, and datetime-related features.

Most columns do not contain missing values. However, source_name and destination_name contain a small number of null values, which will require treatment during data cleaning.

The dataset also contains multiple object-type columns that can later be optimized into categorical and datetime formats for improved memory efficiency and analysis.

In [10]:
df.describe()

,start_scan_to_end_scan,cutoff_factor,actual_distance_to_destination,actual_time,osrm_time,osrm_distance,factor,segment_actual_time,segment_osrm_time,segment_osrm_distance,segment_factor
count,144867.000000,144867.000000,144867.000000,144867.000000,144867.000000,144867.000000,144867.000000,144867.000000,144867.000000,144867.00000,144867.000000
mean,961.262986,232.926567,234.073372,416.927527,213.868272,284.771297,2.120107,36.196111,18.507548,22.82902,2.218368
std,1037.012769,344.755577,344.990009,598.103621,308.011085,421.119294,1.715421,53.571158,14.775960,17.86066,4.847530
min,20.000000,9.000000,9.000045,9.000000,6.000000,9.008200,0.144000,-244.000000,0.000000,0.00000,-23.444444
25%,161.000000,22.000000,23.355874,51.000000,27.000000,29.914700,1.604264,20.000000,11.000000,12.07010,1.347826
50%,449.000000,66.000000,66.126571,132.000000,64.000000,78.525800,1.857143,29.000000,17.000000,23.51300,1.684211
75%,1634.000000,286.000000,286.708875,513.000000,257.000000,343.193250,2.213483,40.000000,22.000000,27.81325,2.250000
max,7898.000000,1927.000000,1927.447705,4532.000000,1686.000000,2326.199100,77.387097,3051.000000,1611.000000,2191.40370,574.250000


**Observation:**

Several numerical variables such as actual_time, osrm_time, and osrm_distance show large standard deviations and wide ranges between minimum and maximum values.

The presence of extremely high maximum values compared to median values indicates potential outliers and right-skewed distributions in operational delivery metrics.

Negative values observed in segment_actual_time and segment_factor may indicate data inconsistencies or exceptional operational cases that require further investigation.

In [11]:
df.isnull().sum()

,0
data,0
trip_creation_time,0
route_schedule_uuid,0
route_type,0
trip_uuid,0
source_center,0
source_name,293
destination_center,0
destination_name,261
od_start_time,0


**Observation:**

Missing values are present only in source_name and destination_name columns, with 293 and 261 missing entries respectively.

Since the proportion of missing values is extremely small compared to the overall dataset size, these values can be treated without causing major information loss.

The remaining columns do not contain any missing values, indicating strong overall data completeness.

In [12]:
df.duplicated().sum()

np.int64(0)

**Observation:**

No duplicate rows were found in the dataset, indicating that each shipment-segment record is uniquely captured.

# Data Type Conversion

Convert datetime columns

In [13]:
datetime_cols = [
    'trip_creation_time',
    'od_start_time',
    'od_end_time'
]

for col in datetime_cols:
    df[col] = pd.to_datetime(df[col])

Convert categorical columns

In [14]:
cat_cols = ['data','route_type']

for col in cat_cols:
    df[col] = df[col].astype('category')

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   data                            144867 non-null  category      
 1   trip_creation_time              144867 non-null  datetime64[ns]
 2   route_schedule_uuid             144867 non-null  object        
 3   route_type                      144867 non-null  category      
 4   trip_uuid                       144867 non-null  object        
 5   source_center                   144867 non-null  object        
 6   source_name                     144574 non-null  object        
 7   destination_center              144867 non-null  object        
 8   destination_name                144606 non-null  object        
 9   od_start_time                   144867 non-null  datetime64[ns]
 10  od_end_time                     144867 non-null  datetim

**Observation:**

Datetime and categorical conversions improve memory efficiency and simplify temporal and categorical analysis in later stages of the project.

Dropping Unknown or Irrelevant Features

In [16]:
unknown_cols = [
    'is_cutoff',
    'cutoff_factor',
    'cutoff_timestamp',
    'factor',
    'segment_factor'
]

df.drop(columns=unknown_cols, inplace=True)

In [18]:
df.shape

(144867, 19)

**Observation:**

Columns with unclear business meaning or limited interpretability were removed to improve analytical clarity and reduce unnecessary complexity.

Missing Value Treatment

In [19]:
df.isnull().sum()

,0
data,0
trip_creation_time,0
route_schedule_uuid,0
route_type,0
trip_uuid,0
source_center,0
source_name,293
destination_center,0
destination_name,261
od_start_time,0


Create mapping for source names

In [20]:
source_mapping = (
    df.groupby('source_center')['source_name']
    .first()
)

Create mapping for destination names

In [21]:
destination_mapping = (
    df.groupby('destination_center')['destination_name']
    .first()
)

Fill missing source names

In [22]:
df['source_name'] = df['source_name'].fillna(
    df['source_center'].map(source_mapping)
)

Fill missing destination names

In [23]:
df['destination_name'] = df['destination_name'].fillna(
    df['destination_center'].map(destination_mapping)
)

Remaining nulls

In [24]:
df['source_name'].fillna('Unknown', inplace=True)

df['destination_name'].fillna('Unknown', inplace=True)

/tmp/ipykernel_4494/1404850758.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['source_name'].fillna('Unknown', inplace=True)
/tmp/ipykernel_4494/1404850758.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try

In [25]:
df.isnull().sum()

,0
data,0
trip_creation_time,0
route_schedule_uuid,0
route_type,0
trip_uuid,0
source_center,0
source_name,0
destination_center,0
destination_name,0
od_start_time,0


**Observation:**

Missing location names were treated using center-based mapping to preserve business consistency and minimize information loss.

# Feature Engineering from Location Data

Extracting Source & Destination of State

In [26]:
df['source_state'] = df['source_name'].str.extract(r'\((.*)\)')

In [27]:
df['destination_state'] = df['destination_name'].str.extract(r'\((.*)\)')

Extracting Source & Destination of City

In [28]:
df['source_city'] = df['source_name'].str.split('_').str[0]

In [29]:
df['destination_city'] = df['destination_name'].str.split('_').str[0]

In [30]:
df[['source_name',
    'source_city',
    'source_state']].head()

,source_name,source_city,source_state
0,Anand_VUNagar_DC (Gujarat),Anand,Gujarat
1,Anand_VUNagar_DC (Gujarat),Anand,Gujarat
2,Anand_VUNagar_DC (Gujarat),Anand,Gujarat
3,Anand_VUNagar_DC (Gujarat),Anand,Gujarat
4,Anand_VUNagar_DC (Gujarat),Anand,Gujarat


**Observation:**

City and state-level features were extracted from shipment location names to enable corridor-level and regional logistics analysis.

# Time-Based Feature Engineering

In [31]:
df['trip_month'] = df['trip_creation_time'].dt.month

In [32]:
df['trip_day'] = df['trip_creation_time'].dt.day

In [33]:
df['trip_hour'] = df['trip_creation_time'].dt.hour

In [34]:
df['trip_weekday'] = (
    df['trip_creation_time']
    .dt.day_name()
)

In [35]:
df[['trip_creation_time',
    'trip_month',
    'trip_day',
    'trip_hour',
    'trip_weekday']].head()

,trip_creation_time,trip_month,trip_day,trip_hour,trip_weekday
0,2018-09-20 02:35:36.476840,9,20,2,Thursday
1,2018-09-20 02:35:36.476840,9,20,2,Thursday
2,2018-09-20 02:35:36.476840,9,20,2,Thursday
3,2018-09-20 02:35:36.476840,9,20,2,Thursday
4,2018-09-20 02:35:36.476840,9,20,2,Thursday


**Observation:**

Temporal features such as month, day, hour, and weekday were extracted to analyze operational delivery patterns across different time periods.

# Trip Duration Feature

In [36]:
df['od_time_diff'] = (
    df['od_end_time'] -
    df['od_start_time']
).dt.total_seconds()/60

In [37]:
df[['od_start_time',
    'od_end_time',
    'od_time_diff']].head()

,od_start_time,od_end_time,od_time_diff
0,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.213637
1,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.213637
2,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.213637
3,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.213637
4,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.213637


**Observation:**

A new feature representing total trip duration in minutes was created using origin-destination timestamps.

# Understanding Shipment Segmentation

Each trip in the dataset is divided into multiple shipment segments. Therefore, multiple rows may belong to the same trip.

To perform meaningful operational analysis, shipment-segment records must be aggregated into trip-level records by combining cumulative metrics while preserving key identifying information.

# Aggregating Shipment Segments

In [38]:
group_cols = [
    'trip_uuid',
    'source_center',
    'destination_center'
]

In [39]:
agg_dict = {

    'actual_time':'sum',
    'osrm_time':'sum',
    'osrm_distance':'sum',
    'segment_actual_time':'sum',
    'segment_osrm_time':'sum',
    'segment_osrm_distance':'sum',

    'trip_creation_time':'first',
    'route_type':'first',
    'source_name':'first',
    'destination_name':'first',
    'od_start_time':'first',
    'od_end_time':'last',
    'start_scan_to_end_scan':'first',
    'od_time_diff':'sum'
}

In [40]:
df_grouped = (
    df.groupby(group_cols)
    .agg(agg_dict)
    .reset_index()
)

In [41]:
df_grouped.shape

(26368, 17)

In [42]:
df_grouped.head()

,trip_uuid,source_center,destination_center,actual_time,osrm_time,osrm_distance,segment_actual_time,segment_osrm_time,segment_osrm_distance,trip_creation_time,route_type,source_name,destination_name,od_start_time,od_end_time,start_scan_to_end_scan,od_time_diff
0,trip-153671041653548748,IND209304AAA,IND000000ACB,6484.0,3464.0,4540.1261,728.0,534.0,670.6205,2018-09-12 00:00:16.535741,FTL,Kanpur_Central_H_6 (Uttar Pradesh),Gurgaon_Bilaspur_HB (Haryana),2018-09-12 16:39:46.858469,2018-09-13 13:40:23.123744,1260.0,22690.879582
1,trip-153671041653548748,IND462022AAA,IND209304AAA,9198.0,4323.0,6037.6386,820.0,474.0,649.8528,2018-09-12 00:00:16.535741,FTL,Bhopal_Trnsport_H (Madhya Pradesh),Kanpur_Central_H_6 (Uttar Pradesh),2018-09-12 00:00:16.535741,2018-09-12 16:39:46.858469,999.0,20989.612955
2,trip-153671042288605164,IND561203AAB,IND562101AAA,96.0,55.0,60.3157,46.0,26.0,28.1995,2018-09-12 00:00:22.886430,Carting,Doddablpur_ChikaDPP_D (Karnataka),Chikblapur_ShntiSgr_D (Karnataka),2018-09-12 02:03:09.655591,2018-09-12 03:01:59.598855,58.0,176.497163
3,trip-153671042288605164,IND572101AAA,IND561203AAB,303.0,155.0,209.1151,95.0,39.0,55.9899,2018-09-12 00:00:22.886430,Carting,Tumkur_Veersagr_I (Karnataka),Doddablpur_ChikaDPP_D (Karnataka),2018-09-12 00:00:22.886430,2018-09-12 02:03:09.655591,122.0,736.676916
4,trip-153671043369099517,IND000000ACB,IND160002AAC,2601.0,1427.0,1975.7409,608.0,231.0,317.7408,2018-09-12 00:00:33.691250,FTL,Gurgaon_Bilaspur_HB (Haryana),Chandigarh_Mehmdpur_H (Punjab),2018-09-14 03:40:17.106733,2018-09-14 17:34:55.442454,834.0,10015.667144


**Observation:**

Shipment-segment records were aggregated into route-level operational records by summing cumulative metrics and preserving identifying trip information.